nasm: bump version to 2.14.02

trungnt2910 opened the pull request:
I've updated `nasm` recipe to version 2.14.03rc2 (latest release).
The update involves a few patches to the source code. 
Included here is the log for `haikuporter nasm`.

[nasm-2.14.03rc2_build.log](https://github.com/haikuports/haikuports/files/4091571/nasm-2.14.03rc2_build.log)

This solves a few problems, such as building dav1d, which I want to attempt next.

trungnt2910 commented:
@korli @waddlesplash
Please review my task, only one hour left and I really want to try building `dav1d` with the `nasm` that I've updated.

Begasus commented on dev-lang/nasm/patches/nasm-2.14.03rc2.patchset:
```diff
@@ -0,0 +1,292 @@
+-infodir		= ${datarootdir}/info
+-datarootdir	= ${prefix}/share
++infodir		= /packages/nasm-2.14.03rc2-1/.self/documentation/info
++datarootdir	= /packages/nasm-2.14.03rc2-1/.self/data
```
Shouldn't be needed, runConfigure should take care of this

trungnt2910 replied:
I forgot to remove that file from my commits. Will remove it soon.

Begasus commented on dev-lang/nasm/nasm-2.14.03rc2.recipe:
```diff
@@ -34,6 +34,8 @@ BUILD_PREREQUIRES="
 	cmd:ld
 	cmd:make
 	cmd:sed
+	cmd:autoreconf
```
Why add autoreconf if you're not using it?
```diff
- 	cmd:autoreconf
```

trungnt2910 replied:
At first, I thought the compile errors was because of config file, as the compiler flags in 14.02 are different, and modifying `config.ac` requires running `autoreconf`.
Now it is not needed, and I will remove it.

korli commented on dev-lang/nasm/patches/nasm-2.14.03rc2.patchset:
```diff
@@ -0,0 +1,292 @@
+Date: Tue, 21 Jan 2020 20:55:56 +0700
+Subject: Fixed various build errors in gcc
+
+Since code in .hpkg packages are only meant to be used for Haiku, I've removed numerous lines of code intended to ensure portablity, but makes Haiku's version of gcc complain.
```
gcc2.95 or gcc 8.3 ?

trungnt2910 replied:
This works for gcc 7.3.0.

trungnt2910 commented:
I've fixed the recipe to remove `cmd:autoreconf`.
Also, I've replaced my old patch with a new and more carefully built `.patchset`.

alaviss commented:
Please switch to `nasm 2.14.02`. We would like to limit the amount unstable software in our repository.

alaviss commented:
The patchset is worrying. Haiku's compiler shouldn't choke on such simple headers, we will have to investigate.

trungnt2910 commented:
I don't know why, right from the start, gcc 7.3.0 on Haiku complains about undefined `bool`, which is already included in `complier.h `. Then it says about faulty function definitions, which are actually macros the nasm developers defined for attributes.

trungnt2910 commented:
I'm creating the recipe for version 2.14.02.
Attached is the log when attempting to build with the unmodified source. 
[nasm-2.14.02_build.log](https://github.com/haikuports/haikuports/files/4099061/nasm-2.14.02_build.log)
Many error messages are duplicates, so here are some common messages:
1: Caused by missing `<stdbool.h>`
> ./include/nasm.h:144:39: error: unknown type name 'bool'; did you mean '_Bool'?
>                         char *special, bool is_norm);
>                                        ^~~~
>                                        _Bool

2: Caused by unrecognized macro for attribute:
> ./include/nasmlib.h:75:20: error: expected declaration specifiers or '...' before numeric constant
>  void * safe_malloc(1) nasm_malloc(size_t);

trungnt2910 commented:
Build for nasm 2.14.02 succeeded on gcc 7.3.0 with the same patch as 2.14.03rc2.
[nasm-2.14.02_build_patched.log](https://github.com/haikuports/haikuports/files/4099203/nasm-2.14.02_build_patched.log)
Recipe will be uploaded in the next commit.

alaviss commented:
OK, so for whatever reason, `"config/config.h"` (generated after configure) wasn't being picked up by nasm.

That file contains everything needed to make nasm works on Haiku (all configuration for `compiler.h` is in there), so the first priority is to figure out why it wasn't picked by gcc.

trungnt2910 commented:
`gcc` does not pick a lot of things.

> #ifdef __HAIKU__ 
> /*  
>  * Declarations from header that the compiler 
>  * fails to recognize on Haiku 
>  */ 
> typedef uint64_t opflags_t; 
> extern const opflags_t nasm_reg_flags[]; 
> #endif
 
These are code that I inserted to one of the header files. It already has `tables.h` included, also, gcc was already called with the `-I./include` switch, which should instruct it to find in the directory containing `tables.h`. However, the compile just simply complains about `opflag_t` being undefined.

alaviss commented:
@trungnt2910 
Investigation complete and it's kinda funny:

So what happens was:
`#include <limits.h>` which `#include <float.h>`, which is fine, except nasm provides a `float.h` under one of its include directory (`asm`), which our `gcc` picks up instead of the gcc-provided one.
By the time `config.h` was actually sourced, a ton of nasm code has been erroneously parsed.

I'd say this is a Haiku bug.

@waddlesplash @pulkomandy any good idea on how to solve this one?

alaviss commented:
Here's the temporary walkaround: http://ix.io/2859. That one patch fix the compilation.

What this does is to lower nasm include priority so Haiku's include will took place first, but this is definitely how this should be fixed.

alaviss commented:
True... and they actually used quoted includes correctly. I guess we can just get convert those `-I` to `-iquote` and do a patch on their side.

trungnt2910 commented:
Build fails on x86_gcc2, while older source (version 2.12.02) works.
Here is the error message:
> /sources/nasm-2.14.03rc2/include/ilog2.h:143: too many `l's in integer constant
It seems that gcc2 does not support `long long`.
So how can I configure the recipe for my secondary architecture?

Begasus commented:
> Build fails on x86_gcc2, while older source (version 2.12.02) works.
> Here is the error message:
> 
> > /sources/nasm-2.14.03rc2/include/ilog2.h:143: too many `l's in integer constant
> 
> It seems that gcc2 does not support `long long`.
> So how can I configure the recipe for my secondary architecture?

Add: SECONDARY_ARCHITECTURES and $secondaryArchSuffix where needed (didn't check this yet)

Begasus commented:
Builds ok for secondary architecture with the patch from @alaviss

Begasus commented:
Maybe remove the 2.14.03rc2 recipe from this PR? (stick to stable release)

trungnt2910 commented:
> /sources/nasm-2.14.03rc2/include/ilog2.h:143: too many `l's in integer constant
> 
> It seems that gcc2 does not support long long.
> So how can I configure the recipe for my secondary architecture?

You mean on every item in the `PROVIDES` and `REQUIRES`?

Begasus commented:
> > /sources/nasm-2.14.03rc2/include/ilog2.h:143: too many `l's in integer constant
> > It seems that gcc2 does not support long long.
> > So how can I configure the recipe for my secondary architecture?
> 
> You mean on every item in the `PROVIDES` and `REQUIRES`?

Check how other recipes use it :) (search for "!x86_gcc2")

trungnt2910 commented:
Adding secondary architecture gives this message:
$secondaryArchSuffix
> warning: invalid preprocessing directive name
It has the same message in many files, then fails.

Begasus commented:
> Adding secondary architecture gives this message:
> $secondaryArchSuffix
> 
> > warning: invalid preprocessing directive name
> > It has the same message in many files, then fails.

Can you push here? Or pastebin the recipe as you have it now?

Begasus commented:
PS, not "everything" needs $secondaryArchSuffix :)

trungnt2910 commented:
x86 does not have vmware_tools created for it, I cannot paste the recipe.
I will commit then revert, if necessary.

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -0,0 +1,59 @@
+	"
+
+BUILD_REQUIRES="
+	haiku_devel$secondaryArchSuffix
```
Correction:
```diff
- 	haiku_devel$secondaryArchSuffix
+ 	haiku${secondaryArchSuffix}_devel
```

trungnt2910 replied:
Fixed this.
So package suffixes always go last?

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -0,0 +1,59 @@
+CHECKSUM_SHA256="34fd26c70a277a9fdd54cb5ecf389badedaf48047b269d1008fbc819b24e80bc"
+PATCHES="nasm-$portVersion.patchset"
+
+ARCHITECTURES="x86_gcc2 x86 x86_64"
```
Mark x86_gcc2 as broken
```diff
- ARCHITECTURES="x86_gcc2 x86 x86_64"
+ ARCHITECTURES="!x86_gcc2 x86 x86_64"
```

trungnt2910 commented:
Current recipe still does not work on x86.

WebPositive does not seem to support attaching files, and vmware_tools does not work in x86, so I'm including a pastebin link.

https://pastebin.com/drhPN3iM

Most of the messages are something like this:
> /sources/nasm-2.14.02/include/compiler.h:166: warning: invalid preprocessing directive name

Begasus commented:
> Current recipe still does not work on x86.
> 
> WebPositive does not seem to support attaching files, and vmware_tools does not work in x86, so I'm including a pastebin [link](https://pastebin.com/drhPN3iM).
> 
> Most of the messages are something like this:
> 
> > /sources/nasm-2.14.02/include/compiler.h:166: warning: invalid preprocessing directive name

For secondary architecture you need to call: haikuporter nasm_x86 (don't add version to it)

trungnt2910 commented:
Ok, build works now, with a few warnings:
> Warning: POLICY WARNING: no matching provides "cmd:nasm_x86" for "bin/nasm-x86"
> Warning: POLICY WARNING: no matching provides "cmd:ndisasm_x86" for "bin/ndisasm-x86"

Like mentioned above, should I add `$secondaryArchSuffix` to these two? If I do, we must invoke using <command-name>-x86.

trungnt2910 commented:
https://pastebin.com/MmvqzQr3
Log here.

Begasus commented:
Looks OK

trungnt2910 commented:
Should we add `$secondaryArchSuffix` to `cmd:nasm` and `cmd:ndisasm` in the PROVIDES section?
If we do, we must call `nasm-x86`.
Is this normal behavior?

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -0,0 +1,59 @@
+
+BUILD()
+{
+	runConfigure ./configure
```
My bad, use "--omit-dirs binDir" etc here to set this to "--bindir=$prefix/bin"

trungnt2910 replied:
How? Can you show this in more detail?

Begasus replied:
Check examples: https://github.com/haikuports/haikuports/blob/dfaebbcdb66af70d8fb5680d7bd52576b24607ad/app-crypt/libu2f_host/libu2f_host-1.1.10.recipe

trungnt2910 replied:
Thanks.

Begasus replied:
"--bindir=$prefix/bin" should be "--bindir=$commandBinDir" as shown in the example

Begasus replied:
https://github.com/haikuports/haikuports/blob/dfaebbcdb66af70d8fb5680d7bd52576b24607ad/app-crypt/libu2f_host/libu2f_host-1.1.10.recipe#L20

Begasus replied:
https://github.com/haikuports/haikuports/blob/dfaebbcdb66af70d8fb5680d7bd52576b24607ad/app-crypt/libu2f_host/libu2f_host-1.1.10.recipe#L69

trungnt2910 replied:
Thanks. It's working now.

trungnt2910 commented:
It doesn't work, gives this error:

> Building ...
> configure: error: expected an absolute directory name for --bindir: 
> Warning: Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 1

Begasus commented:
> It doesn't work, gives this error:
> 
> > Building ...
> > configure: error: expected an absolute directory name for --bindir:
> > Warning: Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 1

Did you included the changes for commandSuffix/commandBinDir?

trungnt2910 commented:
Yes I did.

Begasus commented:
Also, remove nasm-2.14.03rc2.recipe from the PR

Begasus commented:
> Yes I did.

Push, don't have too much time to check on that atm :)

Begasus commented on dev-lang/nasm/nasm-2.14.03rc2.recipe:
```diff
@@ -14,24 +14,32 @@ SOURCE_URI="https://www.nasm.us/pub/nasm/releasebuilds/$portVersion/nasm-$portVe
 PATCHES="nasm-$portVersion.patchset"
 
-ARCHITECTURES="x86_gcc2 x86 x86_64"
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
```
Please remove this recipe, stay with the stable release, not the release candidate

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -17,10 +17,17 @@ PATCHES="nasm-$portVersion.patchset"
-	cmd:nasm = $portVersion compat >= 1
-	cmd:ndisasm = $portVersion compat >= 1
+	cmd:nasm$commandSuffix = $portVersion compat >= 1
+	cmd:ndisasm$commandSuffix = $portVersion compat >= 1
```
No need for compat here

trungnt2910 commented:
Thanks. Installed properly in a clean Haiku_x86_gcc2 machine.

alaviss commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -1,39 +1,48 @@
+	commandBinDir=$prefix/bin
+fi
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
```
```suggestion
ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
```

We can't test this architecture since it's no longer built by Haiku.

alaviss commented on dev-lang/nasm/patches/nasm-2.14.02.patchset:
```diff
@@ -0,0 +1,270 @@
+From 4f2b8256a735d00ef50d76abeb56b772d8f3b2ed Mon Sep 17 00:00:00 2001
```
Please switch to [this] patch instead, which fix the actual root of the problem.

[this]: http://ix.io/289y

trungnt2910 replied:
Thanks. I'll update this later.

trungnt2910 commented:
Ran into an error while building nasm with the new patchset:

> Applying patchset "/boot/home/haikuports/dev-lang/nasm/patches/nasm-2.14.02.patchset" ...
> error: sha1 information is lacking or useless (Makefile.in).
> error: could not build fake ancestor
> hint: Use 'git am --show-current-patch' to see the failed patch
> Command '['git', 'am', '--ignore-whitespace', '-3', '--keep-cr', u'/boot/home/haikuports/dev-lang/nasm/patches/nasm-2.14.02.patchset']' returned non-zero exit status 128

alaviss commented:
Rename it to `.patch` and it should applies. I made that patchset based on `nasm` latest git head so I could upstream it, which made it incompatible with HaikuPorter's `.patchset`.

I could generate a proper patchset later though, but I'll push the changes later as needed.

trungnt2910 commented:
@alaviss So you are Hiếu Lê?

trungnt2910 commented:
It fails after renaming to `.patch`.

Begasus commented:
Diff
```diff
-	cmd:nasm = $portVersion compat >= 1
-	cmd:ndisasm = $portVersion compat >= 1
+	cmd:nasm$commandSuffix = $portVersion compat >= 1
+	cmd:ndisasm$commandSuffix = $portVersion compat >= 1
```

trungnt2910 commented:
> Diff
> -	cmd:nasm = $portVersion compat >= 1
> -	cmd:ndisasm = $portVersion compat >= 1
> +	cmd:nasm$commandSuffix = $portVersion compat >= 1
> +	cmd:ndisasm$commandSuffix = $portVersion compat >= 1

So we **do** need this after all?

Begasus commented:
> > Diff
> > 
> > * cmd:nasm = $portVersion compat >= 1
> > * cmd:ndisasm = $portVersion compat >= 1
> > 
> > 
> > * cmd:nasm$commandSuffix = $portVersion compat >= 1
> > * cmd:ndisasm$commandSuffix = $portVersion compat >= 1
> 
> So we **do** need this after all?

Even if the outcome is the same, it's good practice and used by others before this :)

trungnt2910 commented:
In my last push, I've cleaned up things a bit, removed version 2.14.03rc2 completely from the history of this branch, also, I've created the correct `.patchset` myself, while still keeping Leorize's credits.

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -1,39 +1,48 @@
+fi
+
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
```
Check ordering from the guidelines: https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines#ordering
architectures before commandbindir ...

trungnt2910 replied:
Sorry, put that in the wrong place :(.

alaviss commented:
> @alaviss So you are Hiếu Lê?

Yes, though please refer to me with @alaviss or leorize, whichever that you prefer. I rarely use my name in OSS contrib unless it's required.

trungnt2910 commented:
> es, though please refer to me with @alaviss or leorize, whichever that you prefer. I rarely use my name in OSS contrib unless it's required.

Xin chào

alaviss commented:
:wave: I'd prefer IRC for small talks instead of PRs though.

trungnt2910 commented:
@alaviss
So are there any more fixes I need to do?
I'd like to get everything done before midnight (when the lunar new year comes).
Thanks.

alaviss commented:
@trungnt2910 
> Xin chào

Btw you can find me in `#haiku` if you wanna chat :)

alaviss commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -42,7 +51,8 @@ TEST_REQUIRES="
 
```
Need `defineDebugInfoPackage` since nasm was built with `-g`

trungnt2910 replied:
What do you mean `-g`?

alaviss replied:
If you pay attention to the building process, you'll find that `gcc` was called with `-g` (generate debuginfo) flag.

trungnt2910 replied:
So should we define for `nasm` and `ndisasm` or nasm only?

alaviss replied:
All binaries.

alaviss commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -1,48 +1,62 @@
 	"
 
+defineDebugInfoPackage nasm /
+	$commandBinDir/nasm /
```
Wrong syntax.

alaviss commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -1,48 +1,62 @@
 	cmd:perl
 	"
 
+defineDebugInfoPackage nasm \
```
```suggestion
defineDebugInfoPackage nasm$secondaryArchSuffix \
```

This should be the last change needed :)

trungnt2910 replied:
@alaviss you mean `$commandSuffix`? That's what the package provides.

alaviss replied:
Refer to: https://github.com/haikuports/haikuports/wiki/HaikuPorter-BuildRecipes#definedebuginfopackage

> usage: `defineDebugInfoPackage [ --directory <toDirectory> ] <basePackageName> <path> ... `

`commandSuffix` don't do much here :)

Begasus replied:
Please adjust a @alaviss suggested (did this before too)

Begasus replied:
package provides "nasm$secondaryArchSuffix"

trungnt2910 replied:
Oh, I misread the suggestion. It's the package name, not the binary name :(. Sorry.

Begasus commented on dev-lang/nasm/nasm-2.14.02.recipe:
```diff
@@ -1,48 +1,62 @@
 
+defineDebugInfoPackage nasm \
+	$commandBinDir/nasm$secondaryArchSuffix \
+	$commandBinDir/ndisasm$secondaryArchSuffix
```
Don't add $secondaryArchSuffix here:
```diff
-	$commandBinDir/nasm$secondaryArchSuffix \
-	$commandBinDir/ndisasm$secondaryArchSuffix
+	$commandBinDir/nasm \
+	$commandBinDir/ndisasm
```

Begasus replied:
Nothing provides: cmd:*$secondaryArchSuffix" (don't add $commandSuffix here)

trungnt2910 commented:
Thanks.
